dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/stack/transport/sip/netty/SipConnLink.java

```diff
@@ -0,0 +1,523 @@
+	 * 
+	 * @see com.ibm.ws.sip.stack.transaction.transport.connections.SIPConnectionAdapter#writeComplete()
+	 */
+	public void writeComplete(MessageContext messageContext) {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(tc,  "writeComplete", "entry [" + System.identityHashCode(messageContext) + ']');
+		}
+
+		// TODO should we release a ByteBuf created by Unpooled.buffer
+		/*
+		// get its buffer and release it
+		WsByteBuffer oldBuffer = messageContext.getWsByteBuffer();
+
+		if (oldBuffer != null) {
+			// recycle buffer instance
+			messageContext.setWsByteBuffer(null);
+			oldBuffer.release();
+		}
+		*/
+		synchronized (m_outMessages) {
+			m_sendPending = false;
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				Tr.debug(tc,  "writeComplete", "m_sendPending = " + m_sendPending);
+			}
+			if (m_readError == null) {
+				sendPendingMessages();
+			} else {
+				// the connection is closing down due to a previous read error.
+				IOException readError = m_readError;
+				m_readError = null;
+				connectionError(readError);
+			}
+		}
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(tc,  "writeComplete", "exit");
+		}
+	}
+
+	/**
+	 * sends as many messages as possible from the outbound queue without
+	 * blocking the calling thread
+	 */
+	private void sendPendingMessages() {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(tc, "sendPendingMessages");
+		}
+		
+		logConnection();
+		synchronized (m_outMessages) {
+			try {
+				if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+					Tr.debug(tc,  "sendPendingMessages", "m_sendPending = " + m_sendPending + "m_outMessages.isEmpty() = " + m_outMessages.isEmpty());
+				}
+				while (!m_sendPending && !m_outMessages.isEmpty()) {
+					if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+						Tr.debug(tc,  "sendPendingMessages", "sending pending messages");
+					}
+					MessageContext messageSendingContext = m_outMessages.removeFirst();
+					if (!sendNow(messageSendingContext)) {
+						if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+							Tr.debug(tc,  "sendPendingMessages", "send will complete later");
+						}
+						// send will complete later
+						break;
+					}
+					
+					// send completed immediately, so try to send more
+					messageSendingContext.writeComplete();
+				}
+				if (m_closing && !m_sendPending && m_outMessages.isEmpty()) {
+					// last message sent out - time to close.
+					close();
+				}
+			} catch (IOException e) {
+				if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+					Tr.debug(tc,  "sendPendingMessages", "IOException", e);
+				}
+				connectionError(e);
+			}
+		}
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(tc, "sendPendingMessages");
+		}
+	}
+
+	/**
+	 * called by the channel framework when new data arrives
```

Comment:
by the netty framework, not the channel framework?

Reply:
updated comment